packages/openapi-sampler/src/samplers/StringSampler.ts

```diff
@@ -55,30 +55,60 @@ export class StringSampler implements Sampler {
     const randExp = new RandExp(pattern);
 
     if (min) {
-      // ADHOC: make a probe for regex using min quantifier value
-      // e.g. ^[a]+[b]+$ expect 'ab', ^[a-z]*$ expect ''
+      return this.sampleMinLength(randExp, min, max);
+    }
+
+    randExp.max = max ?? randExp.max;
+    randExp.randInt = (a, b) => Math.floor((a + b) / 2);
+
+    const result = randExp.gen();
+
+    return !!max && result.length > max && this.hasInfiniteQuantifier(pattern)
+      ? this.sampleInfiniteQuantifier(randExp, max)
+      : result;
+  }
+
+  private hasInfiniteQuantifier(pattern: string | RegExp) {
+    const pat = pattern.toString();
+
+    return ['+', '*', ',}'].some((q) => pat.includes(q));
+  }
+
+  private sampleInfiniteQuantifier(randExp: RandExp, max: number): string {
+    randExp.randInt = (a, b) => Math.floor((a + b) / 2);
 
-      randExp.max = 0;
-      randExp.randInt = (a, _) => a;
+    for (let i = 1, lmax = max; lmax > 0; lmax = Math.floor(max / ++i)) {
+      randExp.max = lmax;
 
       const result = randExp.gen();
 
-      if (result.length >= min) {
+      if (result.length <= max) {
         return result;
       }
+    }
 
-      // ADHOC: fallback for failed min quantifier probe with doubled min length
+    return '';
+  }
 
-      randExp.max = 2 * min;
-      randExp.randInt = (a, b) => Math.floor((a + b) / 2);
+  private sampleMinLength(randExp: RandExp, min: number, max: number) {
+    // ADHOC: make a probe for regex using min quantifier value
+    // e.g. ^[a]+[b]+$ expect 'ab', ^[a-z]*$ expect ''
 
-      return this.adjustMaxLength(randExp.gen(), max);
+    randExp.max = 0;
```

Comment:
Try to generate with minimum length, and then if this fails, just double it

```suggestion
    randExp.max = min;
```